### ISSUE TITLE
Add badges for Wasm compatibility

### ISSUE THREAD

**Issue by johnpryan** [dart-lang/pub-dev#6785]
Flutter web plugins are currently using `dart:html` and `package:js` to wrap JavaScript APIs. The new version of Dart-JavaScript interop will depend on `package:web` and `dart:js_interop`, respectively.

This is a proposal to display a "Wasm compatible" badge if the package imports the correct libraries, and "Wasm incompatible" if it is using the old libraries.

cc: @kevmoo @joshualitt

**Comment by IchordeDionysos** [comment]
This should be possible by checking if the package depends on `dart:html`, `dart:js`, `dart:js_util`, or `package:js`?

Or any other dependency that does not support WASM (e.g. because it or one of its transitive dependencies depend on the above), right?

**Comment by kevmoo** [comment]
@IchordeDionysos – yep!

**Comment by IchordeDionysos** [comment]
Hmm, okay, the tricky part is that pub.dev does not know whether the package depends on `dart:xxx` packages ...
https://github.com/flutter/packages/blob/main/packages/video_player/video_player_web/lib/src/video_player.dart#L6

This might have been easily doable (if this was already indexed / available) in the pub.dev database.

**Comment by isoos** [comment]
> Hmm, okay, the tricky part is that pub.dev does not know whether the package depends on dart:xxx packages ...

Not sure if this is a typo or not, so just to be specific about it: pub.dev does package analysis (using `package:pana`) and part of it is to scan the dependency graph for `dart:xxx` library uses. We have started analysis for wasm-compatibility with the recent release, however it is not exposed on the UI yet.

**Comment by IchordeDionysos** [comment]
@isoos Actually, in my assumption, I thought the scan would not detect uses of `dart:xxx`, so wasn't a typo, just my lack of knowledge.

I came to that conclusion as usages of `dart:xxx` packages seem to be not exposed in the UI one pub.dev as well and I thought the reason is the data is not available.

e.g. `video_player_web` uses `dart:html`:
https://github.com/flutter/packages/blob/main/packages/video_player/video_player_web/lib/video_player_web.dart#L6
But it's not listed as a dependency here:
https://pub.dev/packages/video_player_web

Definitely nice to hear support for this badge is progressing 🙌

**Comment by IchordeDionysos** [comment]
> however it is not exposed on the UI yet.
There is no filter yet, but this seems to work already 😍 
https://pub.dev/packages?q=is%3Awasm-ready
At least I can't find the `video_player` nor `video_player_web` package :)

**Comment by isoos** [comment]
> But it's not listed as a dependency here:
> https://pub.dev/packages/video_player_web

We list only package dependencies on the  package page (direct ones in the infobox, direct+transient in the package score tab). It may be useful to also expose the SDK library dependencies (`dart:*`) too, e.g. using `dart:io` may or may not be a concern for users of the package, but we don't have any immediate plan to implement these.

> There is no filter yet, but this seems to work already

We are still evaluating if the analysis is correct and makes sense.

**Comment by isoos** [comment]
> We are still evaluating if the analysis is correct and makes sense.

Also, almost forgot: if you see something wrong with the tagging (either false positive or false negative), please report it (here or in `package:pana`).

**Comment by IchordeDionysos** [comment]
Yeah seen a few where it seemed in correct, will  compile a list.

**Comment by amrgetment** [comment]
@IchordeDionysos This filter shows image_picker and other packages that still don’t support wasm yet or in-progress
https://pub.dev/packages?q=is%3Awasm-ready

image_picker still in-progress 
https://github.com/flutter/flutter/issues/117022

**Comment by amrgetment** [comment]
@isoos I think  package:pana may need to check for dart:html dart:js dart:js_util package:js in the package or its dependencies like @IchordeDionysos said

**Comment by IchordeDionysos** [comment]
Definition:
- false positive: Falsely marked as WASM-ready
- false negative: Falsely marked as non-WASM-ready

**intl: false positive**
The package still seems to use [`dart:html`](https://github.com/search?q=repo%3Adart-lang%2Fi18n+dart%3Ahtml+path%3A%2F%5Epkgs%5C%2Fintl%5C%2Flib%5C%2F%2F&type=code)

**i18n_extension: false positive**
Also, packages depending on intl are listed as supported.

**flutter_keyboard_visibility: false positive**
Depends on [`dart:html`](https://github.com/MisterJimson/flutter_keyboard_visibility/blob/master/flutter_keyboard_visibility_web/lib/flutter_keyboard_visibility_web.dart#L1)

**intercom_flutter: false positive**
Depends on [`dart:js` and `dart:html](https://github.com/v3rm0n/intercom_flutter/blob/master/intercom_flutter_web/lib/intercom_flutter_web.dart#L3-L4)

**oauth2_client: false positive:**
Depends on a [flutter_secure_storage](https://pub.dev/packages/flutter_secure_storage) which depends on [flutter_secure_storage_web](https://pub.dev/packages/flutter_secure_storage_web) which depends on `package:js`

**sensors_plus: false positive:**
Depends on both [`dart:html` and `dart:js_util`](https://github.com/fluttercommunity/plus_plugins/blob/main/packages/sensors_plus/sensors_plus/lib/src/web_sensors.dart#L3-L5)

**uni_links: false positive:**
Depends on [`dart:html`](https://github.com/avioli/uni_links/blob/master/uni_links_web/lib/uni_links_web.dart#L1)

**fetch_client: false negative** ???
There is a pre-release version with WASM support, but still listed as unsupported?
I guess that might be expected due to the pre-release

**Comment by kevmoo** [comment]
Just uploaded https://pub.dev/packages/flutter_markdown – which I *think* is wasm-ready, but it's not being labeled

**Comment by jonasfj** [comment]
Looking at `uni_links` I think we might not be handling federated plugins correctly.

I'm pretty sure that `flutter.plugin.platforms.web.default_package` causes some stuff to be generated, and our import analysis does not take that into account.

But if we were to use `flutter.plugin.platforms` in platform detection we might need reference documentation for `flutter.plugin.platforms`, otherwise it'll be very hard to maintain (or even know if it's correct). This, is perhaps something we should push for.

**Comment by jonasfj** [comment]
I guess technically a federated plugin like `uni_links` can be used with wasm, it's just that you'd have to write your own platform implementation :see_no_evil:

The plugin does not directly import `dart:html` so probably possible to override the default web implementation [uni_links_web](https://pub.dev/packages/uni_links_web). It's probably even documented in the federated plugin docs somewhere.

**Comment by IchordeDionysos** [comment]
The thing is when you research a package you want it to work on all platforms and not have to write your own implementation for it.

**Comment by jonasfj** [comment]
Yeah, but if I have to choose between false-positive / and true-negative, then in the case of platform tagging, false-positive isn't the worst.

But yes, we need to do better at understand flutter plugins.

**Comment by kevmoo** [comment]
Could we start with an FYI on the score tab to show how we're evaluating a package?

Might be a nice tip-toe into this space without affecting scores

**Comment by JaffaKetchup** [comment]
Not sure where the best place to report general false negatives are, maybe pana (eg. https://github.com/dart-lang/pana/issues/1324) would be more suited, but I think until packages can define that they do or do not support Wasm (as is the case with platforms) in their pubspec, then it will be difficult to do this accurately.

For example, ['package:logger'](https://pub.dev/packages/logger) is reported as non-Wasm ready, because one of it's files/internal libraries does a conditional import of 'dart:io' (but the package is not a plugin, and this is the primary library), which means 'package:flutter_map' becomes non-Wasm ready, even though it is proved to build with Wasm.

Until maintainers get a way to report Wasm compatibility themselves (as with platforms), then awarding badges might be an issue. Or, pana needs to get better at platform detection for conditional imports. Or, maybe 'package:logger' is just written incorrectly for pana detection? Not sure.

**Comment by kevmoo** [comment]
This should likely be copied to `pkg:pana`.

Thoughts @jonasfj ?

**Comment by jonasfj** [comment]
Please file again pana.
But really `package:logger` doesn't conditionally import `dart:io`.
https://github.com/SourceHorizon/logger/blob/main/lib/src/outputs/advanced_file_output_stub.dart#L2

**Comment by JaffaKetchup** [comment]
Ah, I see, so the library is written somewhat incorrectly. What does pana do with conditional imports anyway
The main point still stands I guess. I would like to mark flutter_map as Wasm compatible anyway.

**Comment by jonasfj** [comment]
Unlike `platforms` where you can explicitly overwrite automatic detection, no such thing is supported for wasm.
https://dart.dev/tools/pub/pubspec#platforms

@kevmoo, @sigurdm this is probably a bit of a hole. wasm isn't really a platform. I suppose we could allow overridding the detection with something like:
```yaml
platform:
  web:
    wasm: true
```

I don't know, if we're ready to formalize what values goes under `platform.<platform>` yet.

---------------

@JaffaKetchup at the moment by strong suggestion would be to contributor to `package:logger`. They could probably remove the unnecessary import of `dart:io`. Cleaning it up will also help platform detection for other packages using `logger`.

**Comment by JaffaKetchup** [comment]
Yeah, something like that would be what I was looking for 👍

**Comment by josxha** [comment]
Similar issues happen here: 

### False positive
[maplibre_gl](https://pub.dev/packages/maplibre_gl/score) is marked wasm compatible on pub.dev. It doesn't include dart:io but uses [maplibre_gl_web](https://pub.dev/packages/maplibre_gl_web/score) for its web implementation like a lot of other flutter plugins do. The problem is that `maplibre_gl_web` does not use compatible js interop. Therefore the [maplibre_gl] used directly by the user can't run on web with wasm either. 

### False negative
[maplibre](https://pub.dev/packages/maplibre/score) is marked wasm incompatible on pub.dev but runs fine on web with wasm. The web implementation gets registered like documented in the [flutter plugin documentation](https://docs.flutter.dev/packages-and-plugins/developing-packages#plugin-platforms) and how it is done in a newly created flutter plugin project. However, the platform channel implementation is directly imported in the platform interface class. Pana detects this and follows the imports to the dart:io imports but does not detect the registered web implementation.

```console
Package not compatible with runtime wasm
Because:
package:maplibre/maplibre.dart that imports:
package:maplibre/src/map.dart that imports:
package:maplibre/src/platform_interface.dart that imports:
package:maplibre/src/native/platform_native.dart that imports:
package:maplibre/src/native/widget_state.dart that imports:
dart:io
```

The pana can improve a lot in detecting false positives and false negatives but I'm not sure if it will be possible to be 100% accurate.

**Comment by sigurdm** [comment]
> [maplibre_gl](https://pub.dev/packages/maplibre_gl/score) is marked wasm compatible on pub.dev. It doesn't include dart:io but uses [maplibre_gl_web](https://pub.dev/packages/maplibre_gl_web/score) for its web implementation like a lot of other flutter plugins do.

[maplibre_gl](https://pub.dev/packages/maplibre_gl/score) is a federated plugin, and only imports directly from https://pub.dev/packages/maplibre_gl_platform_interface, which is a generic interface for. One could (at least in theory) plug  another wasm-compatible web-implementation in.

We could consider it not wasm-compatible if its default implementation is not wasm-compatible. But that is not the current behavior, and not clear that it is correct. (It would also severely make the platform detection less generic.)



The importing code is generated by:
